universe/auto_syncer.go

```diff
@@ -227,55 +233,178 @@ func (f *FederationEnvoy) syncServerState(ctx context.Context,
+		return fmt.Errorf("cannot push proof unable to connect "+
+			"to remote server(%v): %w", addr.HostStr(), err)
 	}
 
-	log.Infof("Pushing new proof to %v federation members, proof_key=%v",
-		len(fedServers), spew.Sdump(key))
+	_, err = remoteUniverseServer.UpsertProofLeaf(
+		ctx, uniID, key, leaf,
+	)
+	if err != nil {
+		return fmt.Errorf("cannot push proof to remote "+
+			"server(%v): %w", addr.HostStr(), err)
+	}
 
-	ctx, cancel := f.WithCtxQuitNoTimeout()
-	defer cancel()
+	return nil
+}
+
+func (f *FederationEnvoy) pushProofToServerLogged(ctx context.Context,
+	uniID Identifier, key LeafKey, leaf *Leaf, addr ServerAddr) error {
+
+	// Ensure that we have a pending sync log entry for this
+	// leaf and server pair. This will allow us to handle all
+	// pending syncs in the event of a restart or at a different
+	// point in the envoy.
+	_, err := f.cfg.FederationDB.UpsertFederationProofSyncLog(
+		ctx, uniID, key, addr, SyncDirectionPush,
+		ProofSyncStatusPending, true,
+	)
+	if err != nil {
+		log.Warnf("unable to log proof sync as pending: %v",
+			err)
+		return nil
+	}
+
+	remoteUniverseServer, err := f.cfg.NewRemoteRegistrar(addr)
+	if err != nil {
+		log.Warnf("cannot push proof, unable to connect "+
+			"to remote server(%v): %v", addr.HostStr(),
+			err)
+		return nil
+	}
+
+	_, err = remoteUniverseServer.UpsertProofLeaf(
+		ctx, uniID, key, leaf,
+	)
+	if err != nil {
+		log.Warnf("cannot push proof to remote "+
+			"server(%v): %v", addr.HostStr(), err)
+		return nil
+	}
+
+	// We did not encounter an error in our proof push
+	// attempt. Log the proof sync attempt as complete.
+	_, err = f.cfg.FederationDB.UpsertFederationProofSyncLog(
+		ctx, uniID, key, addr, SyncDirectionPush,
+		ProofSyncStatusComplete, false,
+	)
+	if err != nil {
+		log.Warnf("unable to log proof sync attempt: %v",
+			err)
+		return nil
+	}
+
+	return nil
+}
+
+// pushProofToFederation attempts to push out a new proof to the current
+// federation in parallel.
+func (f *FederationEnvoy) pushProofToFederation(ctx context.Context,
+	uniID Identifier, key LeafKey, leaf *Leaf, fedServers []ServerAddr,
+	proofSyncLogged bool) {
+
+	log.Infof("Pushing proof to %v federation members, proof_key=%v",
+		len(fedServers), spew.Sdump(key))
 
 	// To push a new proof out, we'll attempt to dial to the remote
 	// registrar, then will attempt to push the new proof directly to the
 	// register.
 	pushNewProof := func(ctx context.Context, addr ServerAddr) error {
-		remoteUniverseServer, err := f.cfg.NewRemoteRegistrar(addr)
-		if err != nil {
-			log.Warnf("cannot push proof unable to connect "+
-				"to remote server(%v): %v", addr.HostStr(),
-				err)
+		// If we are logging proof sync attempts, we will use the
+		// logged version of the push function.
+		if proofSyncLogged {
+			err := f.pushProofToServerLogged(
+				ctx, uniID, key, leaf, addr,
+			)
+			if err != nil {
+				log.Warnf("cannot push proof via logged "+
+					"server push: %v", err)
+				return nil
```

Comment:
nit: can remove the `return nil` as we're returning just below anyway.

Reply:
I did it this way because it makes it safer for any changes which add code after this `if` clause. I think my IDE collapses `if err != nil` clauses sometimes, which means I might miss the lack of a `return`. What do you think of that idea? Is the extra `return nil` dead code a bother?

Reply:
I've removed `return nil`. Don't think we can return error because of `fn.ParSlice`.